### PR TITLE
Make it extremely easy to boot up Starshot with LANDO

### PR DESCRIPTION
**Pull request by MatthieuScarset**
Following the example in #50 I added a `.lando.yml` file to help users who prefer this tool, making it ridiculously easy to set up Starshot with LANDO. 

This PR adds that to the README too:
```
lando start && lando install && lando login
```

### CODE REVIEW

File: README.md
Comment (phenaproxima) [inline README.md:28]: This is, essentially, _four_ commands separated by &&, which is a bit...twitchy. Is there any way to script this as truly a single command, as I did with `.ddev/commands/host/install`?

Also, if `lando install` is already a thing, maybe we should rename both commands to `install-starshot` for consistency.

File: .lando.yml
Comment (phenaproxima) [inline .lando.yml:24]: I'm not sure we need to change `drush.yml`, since we're injecting `DRUSH_OPTIONS_URI` into the container (I think).

Comment (phenaproxima) [reply .lando.yml:24]: Another thing I realize this is missing -- in the DDEV version, a browser window automatically opens up. Can we get the same thing happening with Lando?

**Comment by phenaproxima** [comment]
Hmmm...I tried to test this PR locally and as soon as I cloned it and ran `lando start`, I hit this:

```
Unhandled rejection TypeError: Cannot read properties of null (reading 'trim')
    at Object.clean (/snapshot/cli/node_modules/@lando/core/node_modules/semver/functions/clean.js)
    at /snapshot/cli/node_modules/@lando/core/utils/is-compatible-version.js
    at /snapshot/cli/node_modules/@lando/core/lib/engine.js
    at /snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js
    at /snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js
    at baseForOwn (/snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js)
    at /snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js
    at baseMap (/snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js)
    at Function.map (/snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js)
    at interceptor (/snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js)
    at Function.thru (/snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js)
    at /snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js
    at arrayReduce (/snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js)
    at baseWrapperValue (/snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js)
    at LodashWrapper.wrapperValue (/snapshot/cli/node_modules/@lando/core/node_modules/lodash/lodash.js)
    at /snapshot/cli/node_modules/@lando/core/lib/engine.js
From previous event:
    at Engine.getCompatibility (/snapshot/cli/node_modules/@lando/core/lib/engine.js)
    at /snapshot/cli/node_modules/@lando/core/hooks/lando-get-compat.js
    at AsyncEvents.<anonymous> (/snapshot/cli/node_modules/@lando/core/index.js)
    at AsyncEvents.handle (/snapshot/cli/node_modules/@lando/core/lib/events.js)
    at /snapshot/cli/node_modules/@lando/core/lib/events.js
From previous event:
    at AsyncEvents.emit (/snapshot/cli/node_modules/@lando/core/lib/events.js)
    at /snapshot/cli/node_modules/@lando/core/lib/lando.js
    at process.processImmediate (node:internal/timers:476:21)
From previous event:
    at Lando.bootstrap (/snapshot/cli/node_modules/@lando/core/lib/lando.js)
    at Object.<anonymous> (/snapshot/cli/bin/lando)
    at Module._compile (pkg/prelude/bootstrap.js:1930:22)
    at Module._extensions..js (node:internal/modules/cjs/loader:1414:10)
    at Module.load (node:internal/modules/cjs/loader:1197:32)
    at Module._load (node:internal/modules/cjs/loader:1013:12)
    at Function.runMain (pkg/prelude/bootstrap.js:1983:12)
    at node:internal/main/run_main_module:28:49

Let's get this party started! Starting app starshot...
ERROR ==> Cannot read properties of null (reading 'trim')
```

Running `lando version` reveals v3.21.0-beta.18. Any ideas?

File: .lando.yml
Comment (MatthieuScarset) [reply .lando.yml]: I needed to change `drush.yml` because this file takes over the variables injected :face_with_diagonal_mouth: 

AFAIK there is no command in Lando to `launch` the website, unfortunately.

I've updated the `.lando.yml` trying to keep it simple.

Comment (phenaproxima) [reply .lando.yml]: > AFAIK there is no command in Lando to launch the website, unfortunately.

Well, it's not that we need _Lando_ to have a command for this -- but for parity with our DDEV script, we need be able to call `open` _on the host machine_ -- that is, from _outside_ the appserver container. DDEV has a way to call commands on the host machine (putting the script in `.ddev/commands/host`). Does Lando have that capability?

File: README.md
Comment (MatthieuScarset) [reply README.md]: Agree `lando install` is not clever.

I can reduce this to 2 commands : `lando start` and `lando drupal:install`

I've updated the README accordingly

File: .lando.yml
Comment (phenaproxima) [reply .lando.yml]: > I needed to change drush.yml because this file takes over the variables injected 🫤

🤔 Hmmm...my understanding, from discussion with the Drush maintainers, is that the environment variables should supersede what's in `drush.yml`. There was a bug around this, but I think they merged it? Maybe it just hasn't made it into a release yet.

Comment (MatthieuScarset) [reply .lando.yml]: Nope, unfortunately this is not possible AFAIK (see [this thread](https://github.com/lando/lando/issues/847#issuecomment-792800913))

**Comment by MatthieuScarset** [comment]
> Hmmm...I tried to test this PR locally and as soon as I cloned it and ran `lando start`, I hit this:

Sorry for this issue. I guess it is caused by some wrong indentation I had in the `.lando.yml` file maybe :shrug: 

I've updated this file and it should be working fine now. 

I've tested on a fresh app with `lando destroy -y` && `lando start && lando drupal:install`

**Comment by Taiger** [comment]
I checked out this branch and ran the lando commands from the readme and everything worked great!
`lando install && lando login`

Thank you for this!

**Comment by AlexSkrypnyk** [comment]
@phenaproxima 
so DDEV was added, now Lando.
There are already issues opened for DDEV not working. 

Is this really necessary to add these unrelated tools to this repo and then support them?

**Comment by Taiger** [comment]
Providing an initial setup for DDEV and Lando is important. Arguably, as important as Drush or Composer tooling.
I was tempted to suggest additional changes for Lando but this initial setup works great as is.

It seems like the aim of of this entire project, is to have better initial defaults. Something that Drupal has struggled with over the years.

**Comment by phenaproxima** [comment]
@AlexSkrypnyk, I frankly agree with @Taiger on this one.

Why not make it as easy as possible for as many people as possible? It doesn't cost us much, and it makes Starshot more accessible. We're not doing _extensive_ support for Lando or DDEV -- just making sure it's easy to spin up. So far, that seems to be working out pretty well, and has received a positive reaction.

I am +1 for this and will merge it once everything is sorted out.

File: .lando.yml
Comment (phenaproxima) [inline .lando.yml:12]: This line is cute but I think we can probably remove it. :)

File: .lando.yml
Comment (phenaproxima) [inline .lando.yml:14]: So, rather than have to pass `-l` or `--uri`, could we do a similar thing to what we're doing with DDEV and inject a `DRUSH_OPTIONS_URI` environment variable into the app server that has the same value as `https://$LANDO_DOMAIN`? That way, Drush will _always_ have the correct URI when run within the app server container and we wouldn't have to remember to pass it.

File: README.md
Comment (phenaproxima) [inline README.md:28]: Let's move the `lando` commands onto their own line.

```suggestion
cd starshot
lando start && lando drupal:install
```

**Review by phenaproxima — CHANGES_REQUESTED** [review]
A couple of very minor changes, and the merge conflicts need to be resolved.

Ideally we would also inject the `DRUSH_OPTIONS_URI` environment variable into the app server container always, so that you never really have to think about passing `--uri`.

**Comment by gitressa** [comment]
@AlexSkrypnyk, I don't understand your opposition to adding DDEV and Lando support. How should developers work on the project, if not with DDEV and Lando, the two most popular developer tools? Also, with Lando and DDEV support out-of-the-box, developers using these tools can get up and running much, much faster and start working on the project, and improving it.

Also, with support for Lando and DDEV, evaluators using these tools can much more easily spin up an instance of Starshot, and check it out.

File: .lando.yml
Comment (MatthieuScarset) [reply .lando.yml:12]: :baby_chick: thank you... I agree BUT... I think emojis makes the experience much better. 

Can we keep it - please ?

File: .lando.yml
Comment (MatthieuScarset) [reply .lando.yml:14]: Yes we can pass the `DRUSH_OPTIONS_URI` in Lando like this: 
```yaml
services:
  appserver:
    overrides:
      environment:
        DRUSH_OPTIONS_URI: "http://mysite.lndo.site"
```

But it does not work because we have a `drush.yml` file committed with a `localhost` value - and it takes precedence over the environment variables (see [previous comment](https://github.com/phenaproxima/starshot-prototype/pull/55#discussion_r1597130590)).

Comment (MatthieuScarset) [reply .lando.yml:14]: @phenaproxima do we _need_ the drush.yml file committed?

If no, then we can use the env var for Drush URI.

If yes, should I add [the `echo` to rewrite it ](https://github.com/phenaproxima/starshot-prototype/pull/55/files/ec0c4ac9c83dcba4d798be2991d1ae0eb9b97a32#diff-307b4548b91e997f23c35d90ed5e9a044e156b90425c9923d6b85d7dc11e67e2R20) again?

**Comment by AlexSkrypnyk** [comment]
@gitressa 
I’m not opposed of adding these great tools. I’m simply asking a question to bring the awareness of the maintenance overhead. If maintainers are okay to take it on themselves - it’s great!

@phenaproxima 
I think your comment answers my question about supporting these tools. Thank you

File: .lando.yml
Comment (AlexSkrypnyk) [reply .lando.yml:12]: A poll we've done in Drupal Slack couple of months ago around using of Emojis in CLI output

Search_-_Drupal_-_Slack

Comment (MatthieuScarset) [reply .lando.yml:12]: So there is a majority of users feeling positive about emojis in CLI :) Thank you for this info @AlexSkrypnyk

Comment (phenaproxima) [reply .lando.yml:12]: No objection to emoji. But let's make it a little briefer. I would suggest:

```
"🚀 Welcome to Starshot! Click on the link below to log in.👇"
```

**Comment by lauriii** [comment]
In order to provide an easy starting point for users who haven't decided on which development environment to use, we have decided to recommend DDEV. You can read https://www.drupal.org/project/ideas/issues/2965681 for context. We should honor that decision here, and instead of providing several options, we should make it clear how users can get started with the recommended local development environment. It makes sense to have a Lando file somewhere that users who are already using Lando can use, but we need to be careful not to do that with a cost to majority of the users.

**Comment by MatthieuScarset** [comment]
> In order to provide an easy starting point for users who haven't decided on which development environment to use, we have decided to recommend DDEV. You can read https://www.drupal.org/project/ideas/issues/2965681 for context. We should honor that decision here, and instead of providing several options, we should make it clear how users can get started with the recommended local development environment. It makes sense to have a Lando file somewhere that users who are already using Lando can use, but we need to be careful not to do that with a cost to majority of the users.

Thank you for pointing out this d.o issue. I didn't know about it. 

As [I commented there](https://www.drupal.org/project/ideas/issues/2965681#comment-15609429), I'm surprised we want to skip 50% of our dev community who currently use Lando. 

Adding a `.lando.yml` file is non-intrusive, very easy to maintain and helps a broader user base. 

I'd like to request a new review to make it extremely easy to boot up Starshot with Lando.

**Comment by phenaproxima** [comment]
@MatthieuScarset, I tried to review this but for some reason, Lando is always giving me this error:

```
ERROR ==> Cannot read properties of null (reading 'trim')
```

...and refusing to go any further, if I run `lando start` having checked out your PR branch. Any advice here? I need to manually test this before I merge it. `lando version` reports v3.21.0-beta.18.

**Comment by MatthieuScarset** [comment]
um... I guess this is something with your local Lando env. 

I've destroyed mine. Removed the folder. Git clone and `lando start && lando drupal:install` and it all worked.

My `lando version` is `v3.21.0-beta.20`

image

**Comment by Manish-Sharma1995** [comment]
It worked for me too without any error. I just cloned the repo and run `lando start && lando drupal:install`. Everything seems to be working perfectly.

My lando version is `v3.18.0`

image

**Comment by phenaproxima** [comment]
Having pondered this for some time, I'm not really convinced we should merge this. My reasons:

* It's taken a lot of iterating and fiddling around to get DDEV support where it is, and adding support for a whole other environment with its own idiosyncrasies means there is more we need to maintain. (Right now, we have to effectively support _two_ development environments: DDEV, and bare metal.)
* DDEV is the officially recommended development environment for Drupal sites. Agree or disagree, that wasn't my decision; Lando is undoubtedly an awesome tool, but it's not the one that seems to be slated for official, wide support in the Drupal community.
* I'd like to keep the support for various development environments as consistent as possible. If we add Lando support now, then maybe we'll add something later for DDEV and then not be able to supply something similar for Lando, due to whatever technical limitations. The maintenance load and added complexity doesn't seem worthwhile to me.

**Pull request closed without merging**
